### PR TITLE
Remove ADUC_ALLOC macros

### PR DESCRIPTION
**Pull request by jw-msft**
- [rmv ADUC_ALLOC macro](https://github.com/Azure/iot-hub-device-update/commit/6d91e477d8bc22cb954d023de8f69f95c9b02723)
- [rmv ADUC_ALLOC_BLOCK macro](https://github.com/Azure/iot-hub-device-update/commit/c2c507978860bc45d714696075a856e2f37b0bbf)
- Add TopicMgmtLifecycle_UninitRetriableOperationContext
- Add CreateCommunicationMgrState and DestroyCommunicationMgrState

### CODE REVIEW

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/agentinfo_request_operation.c [hunk 1]
```diff
@@ -633,22 +645,22 @@ ADUC_Retriable_Operation_Context* CreateAndInitializeAgentInfoRequestOperation()
...
+    context = tmp;
+    tmp = NULL;
 
 done:
 
     free(operationDataContext);
-    free(context);
+    free(tmp);
 
     if (config != NULL && incremented_config_refcount)
```
Comment (Nox-MSFT) [inline src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/agentinfo_request_operation.c:660]: We probably don't need 'incremented_config_refcount' anymore.

File: src/extensions/agent_modules/device_update_workflow/module/src/device_update_workflow_module.c [hunk 2]
```diff
@@ -104,35 +104,38 @@ int DeviceUpdateWorkflowModule_DoWork(ADUC_AGENT_MODULE_HANDLE handle)
...
     workflowData->CommunicationChannel = ADUC_CommunicationChannelType_MQTTBroker;
-    interface->moduleData = workflowData;
-    interface->destroy = DeviceUpdateWorkflowModule_Destroy;
-    interface->getContractInfo = DeviceUpdateWorkflowModule_GetContractInfo;
-    interface->doWork = DeviceUpdateWorkflowModule_DoWork;
-    interface->initializeModule = DeviceUpdateWorkflowModule_Initialize;
-    interface->deinitializeModule = DeviceUpdateWorkflowModule_Deinitialize;
 
-    handle = (ADUC_AGENT_MODULE_HANDLE)interface;
+    tmp->moduleData = workflowData;
```
Comment (Nox-MSFT) [inline src/extensions/agent_modules/device_update_workflow/module/src/device_update_workflow_module.c:126]: Set workflowData to NULL here and free(workflowData) after done: ? (follow your pattern)

File: src/utils/config_utils/src/config_utils.c [hunk 3]
```diff
@@ -249,7 +249,11 @@ bool ADUC_Json_GetAgents(JSON_Value* root_value, size_t* agentCount, ADUC_AgentI
         goto done;
     }
 
-    ADUC_ALLOC_BLOCK(*agents, agents_count, sizeof(ADUC_AgentInfo));
+    *agents = calloc(agents_count, sizeof(ADUC_AgentInfo));
```
Comment (Nox-MSFT) [inline src/utils/config_utils/src/config_utils.c:252]: at line 284, we should pass agents_count insted of *agentCount? (*agentCount wont' get assigned if we goto done at line 255)

Comment (jw-msft) [reply src/utils/config_utils/src/config_utils.c:252]: *agentCount is not set until 6 lines down (so *agentCount here would be 0)--we don't want to set that out param until prove that we can allocate the memory.  The variable names are too similar, so renaming agents_count (Note the underscore char) to `json_array_agents_count` here.

File: src/extensions/agent_modules/device_update_workflow/module/src/device_update_workflow_module.c (hunk 2 above)
Comment (jw-msft) [reply src/extensions/agent_modules/device_update_workflow/module/src/device_update_workflow_module.c:126]: good catch. set to NULL and doing ADUC_WorkflowData_Uninit(workflowData) and free(workflowData) unconditionally after done: label

File: src/utils/config_utils/src/config_utils.c (hunk 3 above)
Comment (jw-msft) [reply src/utils/config_utils/src/config_utils.c]: revamped agents json array parsing based on feedback and made it only return single element (1st element) but log warning if array size is greater than 1

File: src/utils/config_utils/src/config_utils.c [hunk 4]
```diff
@@ -217,74 +227,75 @@ static void ADUC_AgentInfoArray_Free(size_t agentCount, ADUC_AgentInfo* agents)
...
     {
         Log_Error(INVALID_OR_MISSING_FIELD_ERROR_FMT, CONFIG_AGENTS);
         goto done;
     }
 
-    const size_t json_array_agents_count = json_array_get_count(agents_array);
+    json_array_agents_count = json_array_get_count(agents_array);
 
-    if (json_array_agents_count == 0)
+    if (json_array_agents_count != 1)
```
Comment (Nox-MSFT) [inline src/utils/config_utils/src/config_utils.c:268]: Should bail if count == 0

Comment (jw-msft) [reply src/utils/config_utils/src/config_utils.c:274]: good catch. fixed.

**Review by Nox-MSFT — APPROVED** [review]

File: src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c [hunk 5]
```diff
@@ -100,24 +100,38 @@ static ADPS_MQTT_CLIENT_MODULE_STATE* ModuleStateFromModuleHandle(ADUC_AGENT_MOD
...
     moduleState = NULL; // transfer ownership
+
+    // successful, so only now transfer moduleInterface to the output handle.
+    handle = moduleInterface;
     moduleInterface = NULL; // transfer ownership
 
 done:
+    // Always free. If successfully transferred to output handle, then these are NOOPs;
+    // otherwise, the output handle is NULL and these get freed if they were allocated.
     free(moduleState);
```
Comment (JeffMill) [inline src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c:135]: better would be Alloc_AdpsMqttClientModuleState(), Free_AdpsMqttClientModuleState() like we do in the rest of the ADUC code.

Free_ModuleState would free both modulestate and moduleinterface.

We also have a free naming that frees the outer pointer (not just members) as well.

File: src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c [hunk 6]
```diff
@@ -100,24 +100,38 @@ static ADPS_MQTT_CLIENT_MODULE_STATE* ModuleStateFromModuleHandle(ADUC_AGENT_MOD
 ADUC_AGENT_MODULE_HANDLE ADPS_MQTT_Client_Module_Create()
 {
     ADUC_AGENT_MODULE_HANDLE handle = NULL;
+
     ADPS_MQTT_CLIENT_MODULE_STATE* moduleState = NULL;
```
Comment (JeffMill) [inline src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c:104]: What does "ADPS_" stand for?

File: src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c [hunk 7]
```diff
@@ -151,55 +151,107 @@ static inline void Log_Debug_With_Frequency(int freqSeconds, const char* format,
...
     }
 
-    Log_Debug("pendingSubscriptions obj created @ %p", state->pendingSubscriptions);
+    tmp->connectionRetryParams = defaultConnectionRetryParams;
+
+    state = tmp;
+    tmp = NULL;
+
+done:
+    free(tmp);
```
Comment (JeffMill) [inline src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c:190]: similar here, would be nice to see a Alloc_ and Free_MqttCommunicationMgrState()

File: src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c [hunk 8]
```diff
@@ -151,55 +151,107 @@ static inline void Log_Debug_With_Frequency(int freqSeconds, const char* format,
...
-    memset(&state->subscribeTopicInfo, 0, sizeof(state->subscribeTopicInfo));
-    state->connectionRetryParams = defaultConnectionRetryParams;
+    state = CreateCommunicationMgrState();
+    if (state == NULL)
+    {
+        goto done;
+    }
 
-    ADUC_ALLOC(interface);
+    interface = calloc(1, sizeof(*interface));
```
Comment (JeffMill) [inline src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c:232]: why is interface a vector of size 1?  In the future will we have multiple interfaces at the same time, and if so, how does ADUC_AGENT_MODULE_HANDLE know the size of ".interface" ?

Comment (JeffMill) [reply src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c:232]: maybe this is being used for zero-init?

File: src/extensions/agent_modules/device_update_workflow/module/src/device_update_workflow_module.c [hunk 9]
```diff
@@ -104,35 +107,42 @@ int DeviceUpdateWorkflowModule_DoWork(ADUC_AGENT_MODULE_HANDLE handle)
...
+    tmp = calloc(1, sizeof(*tmp));
+    if (tmp == NULL)
+    {
+        goto done;
+    }
 
     // TODO (nox-msft) - pass in launchArgs
-    ret = AzureDeviceUpdateCoreInterface_Create((void**)&workflowData, 0, NULL);
-    if (!ret)
+    if (!AzureDeviceUpdateCoreInterface_Create((void**)&workflowData, 0, NULL))
```
Comment (JeffMill) [inline src/extensions/agent_modules/device_update_workflow/module/src/device_update_workflow_module.c:122]: why the void** cast? Does this method not take a ADUC_WorkflowData** ?

File: src/utils/config_utils/src/config_utils.c [hunk 10]
```diff
@@ -217,70 +227,81 @@ static void ADUC_AgentInfoArray_Free(size_t agentCount, ADUC_AgentInfo* agents)
...
-            goto done;
-        }
+    if (!ADUC_AgentInfo_Init(tmp, cur_agent_obj))
+    {
+        Log_Error("agent info init");
+        goto done;
     }
 
     succeeded = true;
+    *outAgentCount = 1;
```
Comment (JeffMill) [inline src/utils/config_utils/src/config_utils.c:299]: `= json_array_agents_count` in case behavior changes in the future.

**Review by JeffMill — APPROVED** [review]

File: src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c (hunk 8 above)
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c:232]: yes, using for zero-init.

File: src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c (hunk 6 above)
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c:104]: @Nox-MSFT Would you know what the 'A' in ADPS stands for? I'm guessing it's short for either Azure or ADU

File: src/extensions/agent_modules/device_update_workflow/module/src/device_update_workflow_module.c (hunk 9 above)
Comment (jw-msft) [reply src/extensions/agent_modules/device_update_workflow/module/src/device_update_workflow_module.c:122]: No, strangely it takes a formal arg of type (void**) and does not even use the context arg--no idea why!  Removing it...

File: src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c (hunk 5 above)
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c:135]: tracking this and the other comments for another follow-up PR.

**Pull request merged**